pkg/ingester/ingester.go

```diff
@@ -161,110 +143,29 @@ func (i *Ingester) running(ctx context.Context) error {
-		if err != nil {
-			return nil, err
-		}
-		i.instances[tenantID] = inst
-		activeTenantsStats.Set(int64(len(i.instances)))
-	}
-	return inst, nil
-}
-
-func (i *Ingester) getInstanceByID(id string) (*instance, bool) {
-	i.instancesMtx.RLock()
-	defer i.instancesMtx.RUnlock()
-
-	inst, ok := i.instances[id]
-	return inst, ok
-}
-
-// forInstanceUnary executes the given function for the instance with the given tenant ID in the context.
-func forInstanceUnary[T any](ctx context.Context, i *Ingester, f func(*instance) (T, error)) (T, error) {
-	var res T
-	err := i.forInstance(ctx, func(inst *instance) error {
-		r, err := f(inst)
-		if err == nil {
-			res = r
-		}
-		return err
-	})
-	return res, err
-}
-
-// forInstance executes the given function for the instance with the given tenant ID in the context.
-func (i *Ingester) forInstance(ctx context.Context, f func(*instance) error) error {
+func (i *Ingester) Push(ctx context.Context, req *connect.Request[pushv1.PushRequest]) (*connect.Response[pushv1.PushResponse], error) {
 	tenantID, err := tenant.ExtractTenantIDFromContext(ctx)
 	if err != nil {
-		return connect.NewError(connect.CodeInvalidArgument, err)
-	}
-	instance, err := i.GetOrCreateInstance(tenantID)
-	if err != nil {
-		return connect.NewError(connect.CodeInternal, err)
-	}
-	return f(instance)
-}
-
-func (i *Ingester) evictBlock(tenantID string, b ulid.ULID, fn func() error) (err error) {
-	// We lock instances map for writes to ensure that no new instances are
-	// created during the procedure. Otherwise, during initialization, the
-	// new PhlareDB instance may try to load a block that has already been
-	// deleted, or is being deleted.
-	i.instancesMtx.RLock()
-	defer i.instancesMtx.RUnlock()
-	// The map only contains PhlareDB instances that has been initialized since
-	// the process start, therefore there is no guarantee that we will find the
-	// discovered candidate block there. If it is the case, we have to ensure that
-	// the block won't be accessed, before and during deleting it from the disk.
-	var evicted bool
-	if tenantInstance, ok := i.instances[tenantID]; ok {
-		if evicted, err = tenantInstance.Evict(b, fn); err != nil {
-			return fmt.Errorf("failed to evict block %s/%s: %w", tenantID, b, err)
-		}
-	}
-	// If the instance is not found, or the querier is not aware of the block,
-	// and thus the callback has not been invoked, do it now.
-	if !evicted {
-		return fn()
+		return nil, connect.NewError(connect.CodeInvalidArgument, err)
 	}
-	return nil
-}
 
-func (i *Ingester) Push(ctx context.Context, req *connect.Request[pushv1.PushRequest]) (*connect.Response[pushv1.PushResponse], error) {
-	return forInstanceUnary(ctx, i, func(instance *instance) (*connect.Response[pushv1.PushResponse], error) {
+	wait, err := i.segmentWriter.ingest(shardKey(0), func(segment segmentIngest) error {
 		for _, series := range req.Msg.Series {
```

Comment:
I had a quick look at the shard key here and currently a single `pushv1.PushRequest` request can have multiple shard keys because they are passed at the `RawProfileSeries` level (for each `req.Msg.Series`). This is because the distributor groups profiles by ingester and sends a batch and in theory profiles in a batch could be assigned to different shards. 

We will need to consume into separate shards here and respond when all segments are flushed?

Reply:
yep, will do